### ISSUE TITLE
Test issue in semantic segmentation using DL4J

### ISSUE THREAD

**Issue by AbdelmajidB** [deeplearning4j/deeplearning4j#8644]
I have trained a U-net based model on DL4J on medical images. The code above is for testing the model. my objective is to test the model on 34 2d medical images separately and get the segmentation metrics (accuracy, recall…) for each image, so i tried two ways:
1- when i put all the 34 images in the test folder (“testI”) and i run the code which will load all the images and test image by image, i get good segmentation performance for all the images,
2- but when i put just one image that i want to test from the 34 in test folder and i execute the code i get different results in term of accuracy, recall…), that means i get different segmentation results for the same images by using the two approaches.
This is my label generator code: https://gist.github.com/AbdelmajidB/fccb871836746dd4f0d6e32828eadcdf The following code is used test the model image by image: https://gist.github.com/AbdelmajidB/92f5fc3013c81b28c898b64118b6adb5

**Comment by eraly** [comment]
Thanks!

**Comment by AlexDBlack** [comment]
If I understand the problem description correctly, this means that somehow you're getting a different labels array, or maybe different network outputs, depending on how you create the iterators.
Have you been able to narrow it down to which one is the problem?

A test case that creates both data pipelines, compares the labels, and compares the output arrays, and shows a difference, would help us here.

**Comment by AbdelmajidB** [comment]
the problem is that i get different network outputs for the same images using the code above. as i described before i want to get the segmentation evaluation metrics (F1 score, accuracy...) for each image separately. so when i put in test folder an image with ground truth and i execute the code, i get the output segmentation for that image, but when i add a second image and i execute the code i get the output segmentation for each image, but for the first image tested before i get different output segmentation and so on as i add other images, which is not normal. I can't understand the cause of this issue.

**Comment by AbdelmajidB** [comment]
you can find here:
1- The test images: https://drive.google.com/file/d/1nF-Rmqa76F9Wl86YMVwKDs75M9D615eV/view?usp=sharing
2- The pretrained model: https://drive.google.com/file/d/196w4g_ZibH9evPj5HTCB29u25aAxovOC/view?usp=sharing
3- MyPathLabelGenerator :https://gist.github.com/AbdelmajidB/fccb871836746dd4f0d6e32828eadcdf
4- The code for testing: https://gist.github.com/AbdelmajidB/92f5fc3013c81b28c898b64118b6adb5

**Comment by AbdelmajidB** [comment]
Hello @AlexDBlack , please have you any idea about this issue? i uploaded in the last post resources that can you need .

**Comment by AbdelmajidB** [comment]
Hello @AlexDBlack , please have you any idea about this issue? i uploaded in the last post resources that can you need .

**Comment by treo** [comment]
Looking at your code for testing, it looks very much like you are just seeing the randomization in filesplit doing what it is supposed to do. 

If you pass a random instance to file split, you will get the files back in random order.  So if you add files to a directory between runs, the output will be in different orders.

If I'm right, the easiest way to solve that problem is to not pass `rng` in line 18. 

If that doesn't fix the issue, we will need either a self-contained testcase (junit 4) or a self-contained project, that we can just clone and run, so we can see what is going on.

**Comment by AbdelmajidB** [comment]
@treo thank you for your reply. If doesn't fix the issue.

i removed rng from line 18, and i show the path of loaded images in MyPathLabelGenerator to see the results of metrics for each image, so if i put all the 59 images in test folder i get the following results: https://gist.github.com/AbdelmajidB/2d48e945bd81e59a37d461b0eee05572
and if i put just image "1.png" i get this results: https://gist.github.com/AbdelmajidB/868c8259e82b36cc98a2131731551370
so you can see that image "1.png" with the other images it is segmented with good values of F1 score.., but when the image tested alone (in second case) i get bad values.

**Comment by AbdelmajidB** [comment]
@treo @AlexDBlack, please, have you any idea about this issue??

**Comment by treo** [comment]
Can you create a self-contained project that we can run? that would make debugging this a lot easier

**Comment by AbdelmajidB** [comment]
@treo @AlexDBlack you can find here the project that you can run: https://drive.google.com/file/d/16Ov_8Zxj5KDATc78nBXSek3GA07-AYOJ/view?usp=sharing

**Comment by AbdelmajidB** [comment]
@treo @AlexDBlack  please any news about this issue.

**Comment by treo** [comment]
The issue is actually pretty simple: You are using `NormalizerMinMaxScaler` to normalize your data.
This is a scaler that has to be fit to your data, and you do fit it here:
https://gist.github.com/AbdelmajidB/92f5fc3013c81b28c898b64118b6adb5#file-semanticsegmentationload2-java-L29-L31

The problem with the approach you have taken here is that you did a classic data prep mistake:
You fit it to the data you are trying to evaluate. So if there is only a single picture, it fits to that single picture and thereby changes the range of values entirely.

There are two possible solutions for you:
1. You serialize your initial `NormalizerMinMaxScaler` that you've used when training the model.
2. Use `ImagePreProcessingScaler` instead of `NormalizerMinMaxScaler`, as it is a scaler that just assumes an input range of 0 to 255.
